// @veneer/core/dist/scripts/utils/colors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var colors = {
  black: 'black',
  blueDarker: 'blue',
  blueDark: 'blue-dark',
  blue: 'blue-darker',
  blueLight: 'blue-light',
  gray: 'gray',
  grayDark: 'gray-dark',
  grayDarker: 'gray-darker',
  grayLight: 'gray-light',
  grayLighter: 'gray-lighter',
  orangeDark: 'orange',
  orange: 'orange-dark',
  orangeLight: 'orange-light',
  purpleDark: 'purple',
  purple: 'purple-dark',
  purpleLight: 'purple-light',
  yellowDark: 'yellow',
  yellow: 'yellow-dark',
  yellowLight: 'yellow-light',
  white: 'white',
  statusGreen: 'status-green',
  statusGreenDark: 'status-green-dark',
  statusGreenLight: 'status-green-light',
  statusRed: 'status-red',
  statusRedDark: 'status-red-dark',
  statusRedLight: 'status-red-light',
  statusYellow: 'status-yellow',
  statusYellowDark: 'status-yellow-dark',
  statusYellowLight: 'status-yellow-light'
};

colors.array = Object.keys(colors).map(function (k) {
  return colors[k];
});

exports.default = colors;
