// @veneer/table/dist/scripts/tabletitle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _core = require('@veneer/core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TableTitle = function TableTitle(_ref) {
  var column = _ref.column,
      disabled = _ref.disabled,
      label = _ref.label,
      sort = _ref.sort,
      sortable = _ref.sortable,
      onChange = _ref.onChange,
      width = _ref.width;

  var sortIcon = void 0;

  switch (sort) {
    case 'ASC':
      sortIcon = _react2.default.createElement(_core.Icon, { name: 'sort', size: 16, color: 'orange', style: { transform: 'rotate(180deg)' } });
      break;
    case 'DESC':
      sortIcon = _react2.default.createElement(_core.Icon, { name: 'sort', size: 16, color: 'orange' });
      break;
    default:
      sortIcon = '';
      break;
  }

  var className = (0, _classnames2.default)({
    'vn-tableheader__cell': true,
    'vn-tableheader__cell-title': true,
    'vn-tableheader__cell-title--disabled': disabled,
    'vn-tableheader__sort-cell': onChange && sortable,
    'vn-tableheader__sort-cell--disabled': disabled
  });
  var onClick = disabled ? undefined : function () {
    return onChange(column, sort);
  };
  var title = onChange && sortable ? _react2.default.createElement(
    'th',
    {
      onClick: onClick,
      style: width ? { width: '' + width } : undefined,
      className: className
    },
    _react2.default.createElement(
      'div',
      null,
      label,
      sortIcon
    )
  ) : _react2.default.createElement(
    'th',
    { style: width ? { width: '' + width } : undefined, className: className },
    _react2.default.createElement(
      'div',
      null,
      label
    )
  );

  return title;
};

TableTitle.propTypes = {
  column: _propTypes2.default.string,
  disabled: _propTypes2.default.bool,
  label: _propTypes2.default.string.isRequired,
  onChange: _propTypes2.default.func,
  sort: _propTypes2.default.oneOf(['none', 'ASC', 'DESC']),
  sortable: _propTypes2.default.bool
};

TableTitle.defaultProps = {
  column: undefined,
  disabled: false,
  onChange: undefined,
  sort: 'none',
  sortable: true
};

TableTitle.displayName = 'TableTitle';

exports.default = TableTitle;
