// @veneer/table/dist/scripts/tablerowselector.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _core = require('@veneer/core');

require('../css/tablerowselector.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Selector = function Selector(_ref) {
  var isDisplayed = _ref.isDisplayed,
      onSelect = _ref.onSelect,
      rowId = _ref.rowId,
      selected = _ref.selected;

  var className = (0, _classnames3.default)({
    'vn-tablerowselector__checkbox--hidden': isDisplayed === false
  });
  return _react2.default.createElement(_core.Checkbox, {
    name: rowId,
    onClick: function onClick() {
      return onSelect(rowId, !selected);
    },
    checked: selected,
    className: className
  });
};

Selector.propTypes = {
  isDisplayed: _propTypes2.default.bool,
  onSelect: _propTypes2.default.func,
  rowId: _propTypes2.default.string,
  selected: _propTypes2.default.bool
};

Selector.defaultProps = {
  isDisplayed: false,
  onSelect: function onSelect() {},
  rowId: undefined,
  selected: false
};

var Image = function Image(_ref2) {
  var image = _ref2.image,
      isDisplayed = _ref2.isDisplayed,
      rowId = _ref2.rowId;

  var className = (0, _classnames3.default)({
    'vn-tablerowselector__wrapper': true,
    'vn-tablerowselector__wrapper--hidden': isDisplayed === false
  });
  return _react2.default.createElement(
    'div',
    { className: className },
    image.type === 'url' && _react2.default.createElement('img', { src: image.content, alt: rowId, width: 32, height: 32, className: 'vn-tablerowselector__image' }),
    image.type === 'icon' && _react2.default.createElement(_core.Icon, { name: image.content, size: 24, color: 'gray-light', className: 'vn-tablerowselector__icon' }),
    image.type === 'letters' && image.content
  );
};

Image.propTypes = {
  image: _propTypes2.default.shape({
    type: _propTypes2.default.oneOf(['url', 'icon', 'letters']).isRequired,
    content: _propTypes2.default.string
  }),
  isDisplayed: _propTypes2.default.bool,
  rowId: _propTypes2.default.string
};

Image.defaultProps = {
  image: undefined,
  isDisplayed: false,
  rowId: undefined
};

var TableRowSelector = function TableRowSelector(props) {
  var _classnames;

  var hover = props.hover,
      selected = props.selected,
      selectionColumn = props.selectionColumn,
      selectionPosition = props.selectionPosition;

  if (selectionColumn === undefined) {
    return null;
  }

  var isCheckboxDisplayed = selectionColumn === 'checkbox' || selectionColumn === 'image' && hover;

  var className = (0, _classnames3.default)((_classnames = {
    'vn-tablecell': true,
    'vn-tablerowselector': true,
    'vn-tablecell--highlighted': hover === true || selected
  }, _defineProperty(_classnames, 'vn-tablecell--highlighted-' + selectionPosition, true), _defineProperty(_classnames, 'vn-tablerowselector--padded', isCheckboxDisplayed), _classnames));

  return _react2.default.createElement(
    'td',
    { className: className },
    _react2.default.createElement(Selector, _extends({}, props, { isDisplayed: isCheckboxDisplayed })),
    selectionColumn === 'image' && _react2.default.createElement(Image, _extends({}, props, { isDisplayed: !isCheckboxDisplayed }))
  );
};

TableRowSelector.propTypes = {
  hover: _propTypes2.default.bool,
  selected: _propTypes2.default.bool,
  selectionColumn: _propTypes2.default.oneOf(['checkbox', 'image']),
  selectionPosition: _propTypes2.default.oneOf(['lonely', 'first', 'middle', 'last', 'none'])
};

TableRowSelector.defaultProps = {
  hover: false,
  selected: false,
  selectionColumn: undefined,
  selectionPosition: 'none'
};

TableRowSelector.displayName = 'TableRowSelector';

exports.default = TableRowSelector;
