// @veneer/table/dist/scripts/tablerow.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _tablecell = require('./tablecell');

var _tablecell2 = _interopRequireDefault(_tablecell);

var _tablecontent = require('./tablecontent');

var _tablecontent2 = _interopRequireDefault(_tablecontent);

var _tablerowselector = require('./tablerowselector');

var _tablerowselector2 = _interopRequireDefault(_tablerowselector);

require('../css/tablerow.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TableRow = function (_React$Component) {
  _inherits(TableRow, _React$Component);

  function TableRow(props) {
    _classCallCheck(this, TableRow);

    var _this = _possibleConstructorReturn(this, (TableRow.__proto__ || Object.getPrototypeOf(TableRow)).call(this, props));

    _this.onMouseEnter = function () {
      _this.setState({ hover: true });
    };

    _this.onMouseLeave = function () {
      _this.setState({ hover: false });
    };

    _this.state = {
      hover: false
    };
    return _this;
  }

  _createClass(TableRow, [{
    key: 'render',
    value: function render() {
      var _classnames,
          _this2 = this;

      var hover = this.state.hover;
      var _props = this.props,
          columnsOrder = _props.columnsOrder,
          config = _props.config,
          idColumn = _props.idColumn,
          isExpanded = _props.isExpanded,
          isOdd = _props.isOdd,
          rowData = _props.rowData,
          selectionPosition = _props.selectionPosition;


      var className = (0, _classnames3.default)((_classnames = {
        'vn-tablerow': true
      }, _defineProperty(_classnames, 'vn-tablerow--highlighted-' + selectionPosition, selectionPosition !== 'none'), _defineProperty(_classnames, 'vn-tablerow--odd', isOdd), _classnames));

      var configArray = config.map(function (item) {
        return item.column;
      });
      var filteredColumns = columnsOrder.filter(function (columnName) {
        return configArray.indexOf(columnName) !== -1;
      });

      return _react2.default.createElement(
        'tr',
        {
          className: className,
          onMouseEnter: this.onMouseEnter,
          onMouseLeave: this.onMouseLeave
        },
        _react2.default.createElement(_tablerowselector2.default, _extends({
          hover: hover,
          image: rowData.image,
          rowId: rowData[idColumn],
          selected: rowData.selected
        }, this.props)),
        filteredColumns.map(function (columnName) {
          var columnIndex = config.findIndex(function (column) {
            return column.column === columnName;
          });
          var lastVisibleColumnId = filteredColumns.slice(-1)[0];
          var expandable = rowData.details && lastVisibleColumnId === columnName;
          var column = config[columnIndex];

          return _react2.default.createElement(
            _tablecell2.default,
            {
              expandable: expandable,
              hover: hover,
              isExpanded: isExpanded,
              key: rowData[idColumn].toString().concat(column.column.toString()),
              onExpand: _this2.props.onExpand,
              rowId: rowData[idColumn],
              selectionPosition: selectionPosition
            },
            _react2.default.createElement(_tablecontent2.default, {
              actionButton: column.actionButton,
              expandable: expandable,
              rowId: rowData[idColumn],
              type: 'menu',
              value: rowData[column.column]
            })
          );
        })
      );
    }
  }]);

  return TableRow;
}(_react2.default.Component);

TableRow.propTypes = {
  columnsOrder: _propTypes2.default.arrayOf(_propTypes2.default.string).isRequired,
  config: _propTypes2.default.arrayOf(_propTypes2.default.object).isRequired,
  idColumn: _propTypes2.default.string.isRequired,
  isOdd: _propTypes2.default.bool,
  isExpanded: _propTypes2.default.bool,
  onExpand: _propTypes2.default.func,
  rowData: _propTypes2.default.objectOf(_propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.node, _propTypes2.default.object])).isRequired,
  selectionPosition: _propTypes2.default.oneOf(['lonely', 'first', 'middle', 'last', 'none'])
};

TableRow.defaultProps = {
  isOdd: false,
  isExpanded: false,
  onExpand: function onExpand() {},
  selectionPosition: 'none'
};

TableRow.displayName = 'TableRow';

exports.default = TableRow;
