// @veneer/core/dist/scripts/appbar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AppBarRightSection = exports.AppBarLeftSection = exports.AppBar = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _breadcrumb = require('./breadcrumb');

var _breadcrumb2 = _interopRequireDefault(_breadcrumb);

require('../css/appbar.css');

require('../css/layout.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var AppBarLeftSection = function AppBarLeftSection(_ref) {
  var children = _ref.children;
  return _react2.default.createElement(
    'div',
    { className: 'vn-appbar__content vn-appbar__content--left' },
    children
  );
};

AppBarLeftSection.propTypes = {
  children: _propTypes2.default.node.isRequired
};

AppBarLeftSection.displayName = 'AppBarLeftSection';

/*--------------------------------------------------*/

var AppBarRightSection = function AppBarRightSection(_ref2) {
  var children = _ref2.children;
  return _react2.default.createElement(
    'div',
    { className: 'vn-appbar__content vn-appbar__content--right' },
    children
  );
};

AppBarRightSection.propTypes = {
  children: _propTypes2.default.node.isRequired
};

AppBarRightSection.displayName = 'AppBarRightSection';

/*--------------------------------------------------*/

var AppBar = function AppBar(_ref3) {
  var children = _ref3.children,
      theme = _ref3.theme,
      menuEnabled = _ref3.menuEnabled,
      className = _ref3.className,
      style = _ref3.style,
      crumbs = _ref3.crumbs;

  var headerClasses = {
    'vn-appbar': true,
    'vn-appbar__menu--enabled': menuEnabled
  };
  if (className) {
    headerClasses[className] = true;
  }

  headerClasses['vn-appbar--' + theme + '-theme'] = true;

  return _react2.default.createElement(
    'header',
    { className: (0, _classnames2.default)(headerClasses), style: style },
    _react2.default.createElement(
      'div',
      { className: 'vn-appbar__content' },
      children
    ),
    crumbs && _react2.default.createElement(_breadcrumb2.default, { className: 'vn-appbar__breadcrumb--' + theme + '-theme', crumbs: crumbs })
  );
};

AppBar.propTypes = {
  children: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    AppBarLeftSection: AppBarLeftSection,
    AppBarRightSection: AppBarRightSection
  })).isRequired,
  className: _propTypes2.default.string,
  theme: _propTypes2.default.string,
  menuEnabled: _propTypes2.default.bool,
  style: _propTypes2.default.object, // eslint-disable-line react/forbid-prop-types
  crumbs: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    text: _propTypes2.default.string.isRequired,
    url: _propTypes2.default.string
  }))
};

AppBar.defaultProps = {
  className: undefined,
  theme: 'hp',
  menuEnabled: false,
  style: undefined,
  crumbs: undefined
};

AppBar.displayName = 'AppBar';

exports.AppBar = AppBar;
exports.AppBarLeftSection = AppBarLeftSection;
exports.AppBarRightSection = AppBarRightSection;
