// @veneer/core/dist/scripts/breadcrumb.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

require('../css/breadcrumb.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var BreadcrumbItem = function BreadcrumbItem(_ref) {
  var className = _ref.className,
      isLastItem = _ref.isLastItem,
      url = _ref.url,
      text = _ref.text,
      props = _objectWithoutProperties(_ref, ['className', 'isLastItem', 'url', 'text']);

  var isClickable = url && !isLastItem;
  var classes = _defineProperty({
    'vn-breadcrumb__item': true
  }, className, className !== undefined);
  return _react2.default.createElement(
    'li',
    { className: (0, _classnames2.default)(classes) },
    isClickable && _react2.default.createElement(
      'a',
      _extends({ className: 'vn-breadcrumb__link', href: url }, props),
      text
    ),
    !isClickable && text
  );
};

BreadcrumbItem.propTypes = {
  className: _propTypes2.default.string,
  isLastItem: _propTypes2.default.bool,
  text: _propTypes2.default.string.isRequired,
  url: _propTypes2.default.string
};

BreadcrumbItem.defaultProps = {
  className: undefined,
  isLastItem: false,
  url: undefined
};

BreadcrumbItem.displayName = 'BreadcrumbItem';

var Breadcrumb = function Breadcrumb(_ref2) {
  var className = _ref2.className,
      crumbs = _ref2.crumbs,
      props = _objectWithoutProperties(_ref2, ['className', 'crumbs']);

  var items = crumbs.map(function (crumb, index) {
    return _react2.default.createElement(BreadcrumbItem, _extends({
      key: crumb.key || crumb.url || crumb.text
    }, crumb, {
      isLastItem: index === crumbs.length - 1
    }));
  });
  if (crumbs.length > 2) {
    items.unshift(_react2.default.createElement(BreadcrumbItem, { className: 'vn-breadcrumb__item--ellipsis', key: '...', text: '...' }));
  }
  var classes = _defineProperty({
    'vn-breadcrumb': true
  }, className, className !== undefined);
  return _react2.default.createElement(
    'ul',
    _extends({}, props, { className: (0, _classnames2.default)(classes) }),
    items
  );
};

Breadcrumb.propTypes = {
  className: _propTypes2.default.string,
  crumbs: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    key: _propTypes2.default.string,
    text: _propTypes2.default.string.isRequired,
    url: _propTypes2.default.string
  })).isRequired
};

Breadcrumb.defaultProps = {
  className: undefined
};

Breadcrumb.displayName = 'Breadcrumb';

exports.default = Breadcrumb;
