// @veneer/core/dist/scripts/header.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _breadcrumb = require('./breadcrumb');

var _breadcrumb2 = _interopRequireDefault(_breadcrumb);

require('../css/header.css');

require('../css/layout.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Header = function Header(_ref) {
  var _ref$theme = _ref.theme,
      theme = _ref$theme === undefined ? 'hp' : _ref$theme,
      props = _objectWithoutProperties(_ref, ['theme']);

  var breadcrumb = void 0;
  if (props.crumbs) {
    breadcrumb = _react2.default.createElement(
      'div',
      { className: 'vn-content-header__block vn-layout__menu-spacing' },
      _react2.default.createElement(_breadcrumb2.default, { crumbs: props.crumbs })
    );
  }

  var classes = (0, _classnames2.default)('vn-header', 'vn-header--' + theme + '-theme');

  return _react2.default.createElement(
    'header',
    { className: classes },
    _react2.default.createElement(
      'div',
      { className: 'vn-header__block vn-layout__menu-spacing' },
      _react2.default.createElement(
        'div',
        { className: 'vn-header__title-block' },
        _react2.default.createElement(
          'h4',
          { className: 'vn-header__title' },
          props.title
        )
      )
    ),
    breadcrumb
  );
};

Header.propTypes = {
  crumbs: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    text: _propTypes2.default.string.isRequired,
    url: _propTypes2.default.string
  })),
  theme: _propTypes2.default.oneOf(['hp', 'veneer']),
  title: _propTypes2.default.string.isRequired
};

Header.defaultProps = {
  crumbs: [],
  theme: 'hp'
};

Header.displayName = 'Header';

exports.default = Header;
