// @veneer/core/dist/scripts/sidemodal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _icon = require('./icon');

var _icon2 = _interopRequireDefault(_icon);

var _modal = require('./modal');

require('../css/sidemodal.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SideModal = function (_Modal) {
  _inherits(SideModal, _Modal);

  function SideModal() {
    _classCallCheck(this, SideModal);

    return _possibleConstructorReturn(this, (SideModal.__proto__ || Object.getPrototypeOf(SideModal)).apply(this, arguments));
  }

  _createClass(SideModal, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      /* This inerith method is not needed by SideModal */
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          side = _props.side,
          title = _props.title;


      var sideClass = side === 'right' ? 'vn-sidemodal--right' : 'vn-sidemodal--left';

      var sideModalClasses = {
        'vn-sidemodal': true
      };

      sideModalClasses[sideClass] = true;
      sideModalClasses[sideClass + '--open'] = this.state.show;

      return (
        // eslint-disable-next-line jsx-a11y/no-static-element-interactions
        _react2.default.createElement(
          'div',
          { className: (0, _classnames2.default)(sideModalClasses), onClick: _modal.Modal.dialogClick },
          _react2.default.createElement(
            'div',
            { className: 'vn-sidemodal__actions' },
            _react2.default.createElement(
              'a',
              {
                className: 'vn-sidemodal__close',
                onClick: this.closeModal,
                role: 'button',
                tabIndex: 0
              },
              _react2.default.createElement(_icon2.default, {
                name: 'delete',
                color: 'gray',
                size: 16,
                className: 'vn-sidemodal__close-icon'
              })
            )
          ),
          title ? _react2.default.createElement(
            'div',
            { className: 'vn-sidemodal__header' },
            _react2.default.createElement(
              'h4',
              null,
              title
            )
          ) : '',
          _react2.default.createElement(
            'div',
            { className: 'vn-sidemodal__content' },
            _react2.default.createElement(
              'div',
              { className: 'vn-sidemodal__body' },
              children
            )
          )
        )
      );
    }
  }]);

  return SideModal;
}(_modal.Modal);

SideModal.propTypes = {
  children: _propTypes2.default.node,
  dismissOnOverlayClick: _propTypes2.default.bool,
  onClose: _propTypes2.default.func,
  show: _propTypes2.default.bool,
  side: _propTypes2.default.oneOf(['left', 'right']),
  title: _propTypes2.default.string
};

SideModal.defaultProps = {
  children: null,
  dismissOnOverlayClick: false,
  onClose: function onClose() {},
  show: false,
  side: 'right',
  title: null
};

SideModal.displayName = 'SideModal';

exports.default = SideModal;
