// @veneer/core/dist/scripts/sidemenu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SideMenuItem = exports.SideMenu = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames3 = require('classnames');

var _classnames4 = _interopRequireDefault(_classnames3);

var _icon = require('./icon');

var _icon2 = _interopRequireDefault(_icon);

require('../css/sidemenu.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var shortid = require('shortid');

var SideMenu = function (_React$Component) {
  _inherits(SideMenu, _React$Component);

  function SideMenu(props) {
    _classCallCheck(this, SideMenu);

    var _this = _possibleConstructorReturn(this, (SideMenu.__proto__ || Object.getPrototypeOf(SideMenu)).call(this, props));

    _this.handleVisibility = function (event) {
      event.preventDefault();
      _this.setState({
        isExpanded: !_this.state.isExpanded
      });
    };

    _this.state = { isExpanded: props.expanded };
    return _this;
  }

  _createClass(SideMenu, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      this.state = { isExpanded: nextProps.expanded };
    }
  }, {
    key: 'render',
    value: function render() {
      var _classnames;

      var _props = this.props,
          children = _props.children,
          logoImagePath = _props.logoImagePath,
          logoTitle = _props.logoTitle,
          logoUrl = _props.logoUrl,
          theme = _props.theme,
          other = _objectWithoutProperties(_props, ['children', 'logoImagePath', 'logoTitle', 'logoUrl', 'theme']);

      var expanded = other.expanded,
          rootProps = _objectWithoutProperties(other, ['expanded']);

      var isExpanded = this.state.isExpanded;


      var classes = (0, _classnames4.default)((_classnames = {
        'vn-side-menu': true
      }, _defineProperty(_classnames, 'vn-side-menu--' + theme + '-theme', true), _defineProperty(_classnames, 'vn-side-menu--expanded', isExpanded), _classnames));

      return _react2.default.createElement(
        'nav',
        _extends({}, rootProps, { className: classes }),
        _react2.default.createElement('div', { className: 'vn-side-menu__overlay', onClick: this.handleVisibility }),
        _react2.default.createElement(
          'div',
          { className: 'vn-side-menu__menu' },
          _react2.default.createElement(
            'a',
            { className: 'vn-side-menu__logo', href: logoUrl, title: logoTitle },
            _react2.default.createElement('img', { alt: logoTitle, className: 'vn-side-menu__logo-image', src: logoImagePath })
          ),
          _react2.default.createElement(
            'ul',
            null,
            children
          )
        ),
        _react2.default.createElement(
          'a',
          {
            className: 'vn-side-menu__toggle',
            onClick: this.handleVisibility,
            'aria-controls': 'menu',
            'aria-label': 'Navigation',
            role: 'button',
            tabIndex: 0
          },
          _react2.default.createElement(_icon2.default, {
            className: 'vn-side-menu__toggle-icon',
            name: 'hamburguer-menu'
          })
        )
      );
    }
  }]);

  return SideMenu;
}(_react2.default.Component);

SideMenu.propTypes = {
  children: _propTypes2.default.node.isRequired,
  expanded: _propTypes2.default.bool,
  logoImagePath: _propTypes2.default.string,
  logoUrl: _propTypes2.default.string,
  logoTitle: _propTypes2.default.string,
  theme: _propTypes2.default.oneOf(['hp', 'veneer'])
};

SideMenu.defaultProps = {
  expanded: false,
  logoImagePath: null,
  logoUrl: null,
  logoTitle: null,
  theme: 'hp'
};

SideMenu.displayName = 'SideMenu';

var SideMenuItem = function (_React$Component2) {
  _inherits(SideMenuItem, _React$Component2);

  _createClass(SideMenuItem, null, [{
    key: 'isExpanded',
    value: function isExpanded(props) {
      return props.expanded || props.selected || SideMenuItem.hasChildSelected(props.children);
    } // eslint-disable-line react/no-multi-comp

  }]);

  function SideMenuItem(props) {
    _classCallCheck(this, SideMenuItem);

    var _this2 = _possibleConstructorReturn(this, (SideMenuItem.__proto__ || Object.getPrototypeOf(SideMenuItem)).call(this, props));

    _this2.handleClick = function (event) {
      if (!_this2.props.url) {
        event.preventDefault();
        _this2.setState({
          isExpanded: !_this2.state.isExpanded
        });
      }
    };

    _this2.state = { isExpanded: SideMenuItem.isExpanded(props) };
    return _this2;
  }

  _createClass(SideMenuItem, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (this.props.expanded !== nextProps.expanded) {
        this.state = { isExpanded: SideMenuItem.isExpanded(nextProps) };
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          children = _props2.children,
          className = _props2.className,
          label = _props2.label,
          icon = _props2.icon,
          imagePath = _props2.imagePath,
          level = _props2.level,
          selected = _props2.selected,
          url = _props2.url,
          urlTarget = _props2.urlTarget,
          other = _objectWithoutProperties(_props2, ['children', 'className', 'label', 'icon', 'imagePath', 'level', 'selected', 'url', 'urlTarget']);

      var expanded = other.expanded,
          rootProps = _objectWithoutProperties(other, ['expanded']);

      var isExpanded = this.state.isExpanded;


      var classes = (0, _classnames4.default)(_defineProperty({
        'vn-side-menu__item': true,
        'vn-side-menu__item--expanded': isExpanded
      }, className, className !== undefined));

      var linkClasses = (0, _classnames4.default)({
        'vn-side-menu__item-link': true,
        'vn-side-menu__item-link--selected': selected
      });

      var IconClasses = (0, _classnames4.default)({
        'vn-side-menu__item-icon': true,
        'vn-side-menu__item-icon--selected': selected
      });

      var linkStyle = { paddingLeft: (level === 0 ? 10 : level * 30) + 'px' };

      var arrowClasses = (0, _classnames4.default)({
        'vn-side-menu__item-toggle-icon': true,
        'vn-side-menu__item-toggle-icon--expanded': isExpanded
      });

      return _react2.default.createElement(
        'li',
        _extends({}, rootProps, { className: classes }),
        _react2.default.createElement(
          'a',
          {
            className: linkClasses,
            href: url,
            onClick: this.handleClick,
            tabIndex: 0,
            style: linkStyle,
            target: urlTarget
          },
          _react2.default.createElement(
            'div',
            { className: 'vn-side-menu__item-col-icon' },
            imagePath != null && _react2.default.createElement('img', { alt: label, className: 'vn-side-menu__item-img', src: imagePath }),
            imagePath == null && icon != null && _react2.default.createElement(_icon2.default, { className: IconClasses, name: icon })
          ),
          _react2.default.createElement(
            'div',
            { className: 'vn-side-menu__item-col-label' },
            label
          ),
          children != null && _react2.default.createElement(
            'div',
            { className: 'vn-side-menu__item-col-toggle' },
            _react2.default.createElement(_icon2.default, { className: arrowClasses, name: 'arrow' })
          )
        ),
        children != null && _react2.default.createElement(
          'ul',
          null,
          _react2.default.Children.toArray(children).map(function (item) {
            return _react2.default.createElement(
              SideMenuItem,
              _extends({}, item.props, {
                key: shortid.generate(),
                level: level + 1
              }),
              item.props.children
            );
          })
        )
      );
    }
  }]);

  return SideMenuItem;
}(_react2.default.Component);

SideMenuItem.hasChildSelected = function (children) {
  return _react2.default.Children.toArray(children).some(function (child) {
    if (!child || !child.props) return false;
    if (child.props.selected && child.props.selected === true) {
      return true;
    }
    return SideMenuItem.hasChildSelected(child.props.children);
  });
};

SideMenuItem.propTypes = {
  children: _propTypes2.default.node,
  className: _propTypes2.default.string,
  expanded: _propTypes2.default.bool,
  icon: _propTypes2.default.string,
  imagePath: _propTypes2.default.string,
  label: _propTypes2.default.string.isRequired,
  level: _propTypes2.default.number,
  selected: _propTypes2.default.bool,
  url: _propTypes2.default.string,
  urlTarget: _propTypes2.default.string
};

SideMenuItem.defaultProps = {
  children: null,
  className: undefined,
  expanded: false,
  icon: null,
  imagePath: null,
  level: 0,
  selected: false,
  url: null,
  urlTarget: '_self'
};

SideMenuItem.displayName = 'SideMenuItem';

exports.SideMenu = SideMenu;
exports.SideMenuItem = SideMenuItem;
