// @veneer/core/dist/scripts/strength_meter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _colors = require('./utils/colors');

var _colors2 = _interopRequireDefault(_colors);

require('../css/strength_meter.css');

var _icon = require('./icon');

var _icon2 = _interopRequireDefault(_icon);

var _tooltip = require('./tooltip');

var _tooltip2 = _interopRequireDefault(_tooltip);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var defaultColor = _colors2.default.gray;

var StrengthMeter = function StrengthMeter(_ref) {
  var title = _ref.title,
      color = _ref.color,
      value = _ref.value,
      tooltipContent = _ref.tooltipContent,
      tooltipAlignment = _ref.tooltipAlignment,
      tooltipPosition = _ref.tooltipPosition;

  var selectedColor = _colors2.default.array.indexOf(color) !== -1 ? color : defaultColor;
  var hasTooltip = tooltipContent !== '';

  var fixPercentage = function fixPercentage(percent) {
    if (percent < 0) {
      return 0;
    } else if (percent > 100) {
      return 100;
    }

    return percent;
  };

  var percentage = fixPercentage(value);

  return _react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(
      'div',
      { className: 'vn-strength-meter__header' },
      _react2.default.createElement(
        'div',
        { className: 'vn-strength-meter__header__title vn-color--' + selectedColor },
        title
      ),
      hasTooltip && _react2.default.createElement(
        'div',
        { className: 'vn-strength-meter__header__right' },
        _react2.default.createElement(
          _tooltip2.default,
          {
            content: tooltipContent,
            position: tooltipPosition,
            alignment: tooltipAlignment
          },
          _react2.default.createElement(_icon2.default, { className: 'vn-strength-meter__header__right__icon', name: 'alerts', color: selectedColor, size: 16 })
        )
      )
    ),
    _react2.default.createElement(
      'div',
      { className: 'vn-strength-meter' },
      _react2.default.createElement('div', {
        className: 'vn-strength-meter__bar vn-background-color--' + selectedColor,
        style: { width: percentage + '%' }
      }),
      _react2.default.createElement('div', { className: 'vn-strength-meter__background-bar vn-background-color--gray' })
    )
  );
};

StrengthMeter.propTypes = {
  title: _propTypes2.default.string,
  color: _propTypes2.default.oneOf(_colors2.default.array),
  value: _propTypes2.default.number,
  tooltipContent: _propTypes2.default.node,
  tooltipAlignment: _propTypes2.default.oneOf(['top', 'right', 'bottom', 'left', 'middle']),
  tooltipPosition: _propTypes2.default.oneOf(['up', 'right', 'down', 'left'])
};

StrengthMeter.defaultProps = {
  title: '',
  color: defaultColor,
  value: 0,
  tooltipContent: '',
  tooltipAlignment: 'right',
  tooltipPosition: 'up'
};

StrengthMeter.displayName = 'StrengthMeter';

exports.default = StrengthMeter;
